addons/udes_stock_refactoring/models/stock_move.py

```diff
@@ -0,0 +1,355 @@
+# -*- coding: utf-8 -*-
+
+from odoo import api, models, fields, _
+from odoo.exceptions import UserError
+
+from .refactor import REFACTOR_REGISTRY
+
+import logging
+
+_logger = logging.getLogger(__name__)
+
+
+# Map move state to the refactor stage.
+U_STOCK_REFACTOR_STAGES = {
+    "confirmed": "confirm",
+    "waiting": "confirm",
+    "assigned": "assign",
+    "partially_available": "assign",
+    "done": "validate",
+}
+
+
+class StockMove(models.Model):
+    _inherit = "stock.move"
+
+    def _get_refactor_class(self, picking_type, stage):
+        """
+        Return the relevant refactor class from the provided 
+        picking type and stage, if applicable.
+        """
+        try:
+            refactor_action = False
+
+            if stage == "confirm":
+                refactor_action = picking_type.u_post_confirm_action
+            elif stage == "assign":
+                refactor_action = picking_type.u_post_assign_action
+            elif stage == "validate":
+                refactor_action = picking_type.u_post_validate_action
+
+            if refactor_action:
+                refactor_class = REFACTOR_REGISTRY[refactor_action]
+                return refactor_class(self.env)
+
+            return refactor_action
+        except KeyError:
+            raise ValueError(_("Refactor action '%s' could not be found." % refactor_action))
+
+    def compute_grouping_key(self):
+        """Compute the grouping key from the picking type move key format"""
+        # TODO MTC: This can be refactored and abstracted at some point with
+        # this with the equivalent for move_line_key
+        # we need to think of how we want to do it
+
+        # The environment must include {'compute_key': True}
+        # to allow the keys to be computed.
+        if not self.env.context.get("compute_key", False):
+            return
+        for move in self:
+            move_vals = {
+                field_name: move[field_name]
```

Comment:
Feel's like there should be a way of figuring out which fields we need for for the key then storing them 

then using them to power a lambda or only recalculating on changes to them, haven't thought about this enough though perhaps we should take a note to do an investigation to see if improves performance.

Reply:
Have raised issue #10656 to track this so it can be discussed more and an investigation sorted.

Will also add a `TODO` in this function.

Reply:
Have added `TODO` for `compute_grouping_key` on Stock Move and Stock Move Line.